### PR TITLE
(DOCSP-26779): Rules v2

### PR DESCRIPTION
**Pull request by cbush**
## Pull Request Info

See https://github.com/mongodb/docs-realm/pull/2581 for docs-realm corresponding changes

### Jira

- https://jira.mongodb.org/browse/DOCSP-26779

### Staged Changes

- https://docs-atlas-staging.mongodb.com/atlas-app-services/docsworker-xlarge/permissions-v2/

### Feature Branch Progress

- [x] manage-apps/configure/config/data_sources
- [x] manage-apps/secure
- [x] sync/get-started
- [x] sync/app-builder/flexible-sync-permissions-guide
- [x] sync/data-model/create-a-schema
- [x] sync/configure/enable-sync
- [x] sync/configure/sync-settings
- [x] sync/configure/permissions
- [x] sync/error-handling/errors
- [x] sync/go-to-production/optimize-sync-atlas-usage
- [x] rules
- [x] rules/roles
- [x] rules/filters
- [x] rules/expressions
- [x] rules/examples
- [x] mongodb/define-roles-and-permissions
- [x] REDIRECTS

### CODE REVIEW

File: source/mongodb/tables/role-configuration.rst
Comment (jonatlee) [inline source/mongodb/tables/role-configuration.rst:49]: I think there is a subtle detail missing here:
- When a request comes in, App Services looks for a rule matching the namespace (ie namespace rule) before looking at roles. If there is no namespace rule corresponding to the namespace of the request, then that's when we'll use the default rule's roles. If there is no matching namespace rule, AND no default rule, I believe the behavior is to deny the request
- Once the set of roles is established, they are checked in role order https://www.mongodb.com/docs/atlas/app-services/rules/roles/#role-order. If no role applies, then the behavior is to deny the request

**Review by jonatlee — COMMENTED** [review]
Haven't looked at everything in excruciating detail yet, but overall looks nice! Excited for the new docs pages + diagrams

File: source/rules/roles.txt
Comment (krollins-mdb) [inline source/rules/roles.txt:228]: I'm curious about the wording change here. "use" doesn't seem accurate, but "play" seems odd too. I've typically seen users and roles described in a couple of ways:
- the user is the role: a user may **be** any of those roles
- the user has a role: a user may **have** any of those roles

File: source/rules/roles.txt
Comment (krollins-mdb) [inline source/rules/roles.txt:187]: This section only deals with Flexible Sync, correct? The Partition-Based Sync stuff is covered elsewhere.

I think we could be clearer here by replacing most instances of "Device Sync" with "Flexible Sync". This makes it clearer that this doesn't work for Partition-Based Sync.

File: source/rules/roles.txt
Comment (krollins-mdb) [inline source/rules/roles.txt:270]: Specifying Flexible Sync in the second sentence, but not in the first feels odd.
```suggestion
When using Device Sync, App Services assigns a role at the beginning of each
Flexible Sync session. The role determines which permissions apply for the duration of a
session.
```

File: source/rules/roles.txt
Comment (krollins-mdb) [inline source/rules/roles.txt:361]: Let's keep these steps parallel by removing the infinitive. Creating a new role is the implied goal of this procedure. If it's not clear enough, we should add a procedure title above all the steps.

We also probably want to avoid the "start from scratch" idiom, as it may be harder for non-native English speakers to understand.

```suggestion
      #. Click :guilabel:`Add role` to define a new role or or use a 
         preset role as a starting point.
```

File: source/rules/roles.txt
Comment (krollins-mdb) [inline source/rules/roles.txt:366]: It's typically a good idea to avoid Latin terms and abbreviations to improve readability.
```suggestion
         that describe the user and/or their relationship to the data. For 
         example, ``Admin`` or ``Owner``.
```

File: source/rules/roles.txt
Comment (krollins-mdb) [inline source/rules/roles.txt:383]: Slight semantic change to make the connection to collection clearer. Otherwise, it may seem like the roles themselves have a role order.
```suggestion
         the :ref:`role order <role-order>` by clicking the arrows on
```

File: source/rules/sync-compatibility.txt
Comment (krollins-mdb) [inline source/rules/sync-compatibility.txt:80]: Not all readers will understand the term "caveat". How do you feel about these alternatives?
- exception
- condition
- consideration

Comment (krollins-mdb) [reply source/rules/sync-compatibility.txt:80]: Same applies to the other mention of caveat.

File: source/rules/sync-compatibility.txt
Comment (krollins-mdb) [inline source/rules/sync-compatibility.txt:162]:
```suggestion
- Adding a new collection to an App Services Schema and defining permissions for
```

**Review by krollins-mdb — APPROVED** [review]
I have some mostly minor suggestions. Pending these and Jonathan's comments, this is looking really good! Thanks for taking this on. Permissions are always hard to sort out.

File: source/mongodb/tables/role-configuration.rst
Comment (cbush) [reply source/mongodb/tables/role-configuration.rst:49]: Namespace ~= collection, or something else entirely?

Comment (jonatlee) [reply source/mongodb/tables/role-configuration.rst:49]: A namespace is composed of a database and collection name; I think the terms "namespace rule" and "collection rule" are used interchangeably internally, but I think we refer to it as "Collection rules" to customers (at least in the Rules UI).

Comment (cbush) [reply source/mongodb/tables/role-configuration.rst]: Cool, that's what we've been saying in the docs as well. I'll stick with that for now, though it might be worth clarifying somewhere that we mean collection within a specific database (if it's not obvious already)

File: source/rules/roles.txt
Comment (cbush) [reply source/rules/roles.txt:228]: Yes! I inherited this verbiage, and I fixed it, and it seems my fix was undone in a merge. I will re-apply.

File: source/rules/roles.txt
Comment (cbush) [reply source/rules/roles.txt:187]: It's tricky because I'm trying to subtly make "Device Sync" mean "Flexible Sync", as it should eventually supersede PBS.

Comment (krollins-mdb) [reply source/rules/roles.txt:187]: Ah, I see. I think we run the risk of confusing some people, but the number is probably pretty small. Might be worth it to gradually shift Device Sync perception the direction you're going.

File: source/mongodb/tables/role-configuration.rst
Comment (krollins-mdb) [inline source/mongodb/tables/role-configuration.rst:53]: I love this summary. It's the most concise and clear explanation of how  roles are evaluated I've seen. 🎉

File: source/mongodb/tables/role-configuration.rst
Comment (cbush) [reply source/mongodb/tables/role-configuration.rst]: No, we already use the term "Sync Mode". Rationale for not just saying "Flexible Sync" is that I'd like to sunset the notion that mode is a first-class differentiators of "Device Sync" (with an eye towards eventually just not talking about PBS).

File: source/rules/expressions.txt
Comment (cbush) [reply source/rules/expressions.txt:34]: I think because I'm referencing it explicitly above, I'm trying to make it as easy on the eyes as possible.

File: source/rules/sync-compatibility.txt
Comment (cbush) [reply source/rules/sync-compatibility.txt:123]: Good catch, fixed!

File: source/mongodb/tables/role-configuration.rst
Comment (cbush) [reply source/mongodb/tables/role-configuration.rst:190]: Rewrap doesn't behave properly without it!

File: source/mongodb/tables/role-configuration.rst
Comment (cbush) [reply source/mongodb/tables/role-configuration.rst:117]: Good call, thanks

File: source/mongodb/tables/role-configuration.rst
Comment (cbush) [reply source/mongodb/tables/role-configuration.rst:104]: Good catch, thanks

File: source/mongodb/tables/role-configuration.rst
Comment (jonatlee) [reply source/mongodb/tables/role-configuration.rst:71]: That is correct. Also, should we include those diagrams that Caleb has been working on somewhere?

File: source/mongodb/tables/role-configuration.rst
Comment (jonatlee) [reply source/mongodb/tables/role-configuration.rst:77]: That looks right to me, I think there are probably two more columns worth expanding that table to though:

| | document_filters is undefined | document_filters.read is undefined, document_filters.write is defined | document_filters.read is defined, document_filters.write is undefined | document_filters.read and document_filters.write are both defined |
| - | -  | - | - | - |
| flex sync enabled | sync incompatible (deny access)  | sync incompatible (deny access) | sync incompatible (deny access) | if `document_filters.write` evaluates to true, evaluate read/write perms. if `document_filters.write` evaluates to false, deny write access, and evaluate `document_filters.read` - if `document_filters.read` evaluates to true, evaluate read perms |
| flex sync disabled | evaluate read/write perms | if `document_filters.write` evaluates to true, evaluate read/write perms; otherwise, deny read/write access | if `document_filters.read` evaluates to true, evaluate read perms; deny writes  | same as above |  

By `evaluate read/write perms` in the table above, I mean look at top-level `read`/ top-level `write`, and the corresponding field-level perms if top-level `read`/`write` are omitted (ie, the existing role permission evaluation)

Comment (jonatlee) [reply source/mongodb/tables/role-configuration.rst:77]: I think the diagrams that Caleb has been working on will help communicate this better!

File: source/mongodb/tables/role-configuration.rst
Comment (jonatlee) [inline source/mongodb/tables/role-configuration.rst:173]: We actually don't perform any checks involving `write` / field-level `write` perms on deleting a document (see https://docs.google.com/document/d/1tK_vKysOVEb5VdSPBxu-0NXB_WZZ5TnjA0fNx7HpKZQ/edit#heading=h.xnyhe9mxtwef for more info)

Note we do however perform the document_filters checks before checking inserts/deletes

File: source/sync/configure/sync-settings.txt
Comment (jonatlee) [inline source/sync/configure/sync-settings.txt:273]: The field will continue to be used in unmigrated apps, however; On second thought, I think it might be worth holding off from making this change until all apps are migrated (to make sure people aren't accidentally deleting their entire permissions configuration if their app isn't migrated yet)

File: source/mongodb/tables/role-configuration.rst
Comment (cbush) [reply source/mongodb/tables/role-configuration.rst:77]: Tried reworking, have a look (when uploaded)

File: source/rules/sync-compatibility.txt
Comment (cbush) [reply source/rules/sync-compatibility.txt:38]: Went back and forth on this myself, updated

File: source/rules/sync-compatibility.txt
Comment (cbush) [reply source/rules/sync-compatibility.txt:35]: Yeah, good catch. This should say top-level read and write instead of "document filter"!

Comment (cbush) [reply source/rules/sync-compatibility.txt:35]: Wait, no, scratch that

File: source/mongodb/tables/role-configuration.rst
Comment (cbush) [reply source/mongodb/tables/role-configuration.rst:27]: Yeah, just a cut & paste issue

File: source/rules/sync-compatibility.txt
Comment (jonatlee) [reply source/rules/sync-compatibility.txt]: `search` isn't used in any capacity in device sync. top-level `read` and `write` do describe the "read-all-fields"/"write-all-fields" behavior in the context of field-level permissions, and have to be booleans. Might be worth clarifying that under the 3rd bullet point?

File: source/mongodb/tables/role-configuration.rst
Comment (jonatlee) [reply source/mongodb/tables/role-configuration.rst:177]: Correction: just double checked and we actually do the document_filters checks *after* checking inserts/deletes.

Comment (jonatlee) [reply source/mongodb/tables/role-configuration.rst:177]: We should probably also update the diagrams as well to capture that delete permission checks are a bit different (they don't check to see if the fields are writable before actually deleting the document)

File: source/sync/app-builder/device-sync-permissions-guide.txt
Comment (jonatlee) [inline source/sync/app-builder/device-sync-permissions-guide.txt:246]: I think we need top-level `"read": true` and `"write": true` in the admin role here as well, or else no fields will have read/write access

File: source/sync/configure/enable-sync.txt
Comment (jonatlee) [inline source/sync/configure/enable-sync.txt:50]: Not relevant to rules, but as an FYI queryable fields being required is actually changing right now (this is what the new UI looks like):

Screen Shot 2023-02-10 at 1 47 50 PM

The tl;dr is queryable fields are becoming an optional step for flexible sync; instead incoming client queries will automatically have fields referenced in the query added to the sync metadata behind-the-scenes. See https://jira.mongodb.org/browse/BAAS-18399 for more information (cc @sean-brandenburg)

Comment (sean-brandenburg) [reply source/sync/configure/enable-sync.txt:50]: Thanks for calling this out. I've also filed [DOCSP-27844](https://jira.mongodb.org/browse/DOCSP-27844) for updating the relevant documentation for this change

Comment (jonatlee) [reply source/sync/configure/enable-sync.txt:50]: Oh cool. Feel free to ignore this then

File: source/rules/sync-compatibility.txt
Comment (jonatlee) [inline source/rules/sync-compatibility.txt:154]: I think this should probably be "updated the Data Source configuration" instead or something similar, since that's where rules for flexible sync will be read from moving forward

File: source/sync/configure/sync-settings.txt
Comment (jonatlee) [reply source/sync/configure/sync-settings.txt:377]: (feel free to resolve this thread)

File: source/rules/roles.txt
Comment (jonatlee) [reply source/rules/roles.txt:316]: (feel free to resolve this thread)

File: source/rules/roles.txt
Comment (jonatlee) [reply source/rules/roles.txt:377]: (feel free to resolve this thread)

File: source/mongodb/tables/role-configuration.rst
Comment (jonatlee) [reply source/mongodb/tables/role-configuration.rst]: (feel free to resolve this thread)

File: source/sync/app-builder/device-sync-permissions-guide.txt
Comment (jonatlee) [reply source/sync/app-builder/device-sync-permissions-guide.txt]: (feel free to resolve this)

File: source/sync/configure/enable-sync.txt
Comment (jonatlee) [reply source/sync/configure/enable-sync.txt:50]: (feel free to resolve this)

**Review by jonatlee — APPROVED** [review]
Overall looks good to me, great job with this 🎉! There's a few open comment threads I opened that I think still should be discussed at least (I left a comment on the threads that can be resolved - I don't think I have permission to resolve them myself 😄).

File: source/mongodb/tables/role-configuration.rst
Comment (cbush) [reply source/mongodb/tables/role-configuration.rst]: I reworked this completely - realizing we don't fall through to default roles, but it's either or. Check it out and let me know.

Comment (jonatlee) [reply source/mongodb/tables/role-configuration.rst]: Looks better! I know I'm probably being overly pedantic here, but I think _technically_
```
Potential roles are any roles
specified for the given collection or, if no roles are specified for the
given collection, the default role(s)
```
would not apply to the example I mentioned above (ie, even though there are no roles in the namespace rule, the default roles do not apply because the namespace rule exists). 

If we're trying to avoid referencing "rule" in the snippet here - I get it, and would be fine with the existing language, but maybe we could add a note somewhere about this odd edge case / behavior?

Comment (cbush) [reply source/mongodb/tables/role-configuration.rst]: I see. We haven't really defined rule as the file that you put roles in. Let's see if I can continue to avoid it.

File: source/sync/configure/sync-settings.txt
Comment (cbush) [reply source/sync/configure/sync-settings.txt:279]: Do you have a link to an announcement or post about the migration? It might be handy to put a reference here in this case.

Comment (jonatlee) [reply source/sync/configure/sync-settings.txt]: One thing worth calling out there is that we're planning on exposing the feature _only for new apps_ tomorrow. We'll likely let it bake for roughly a week before attempting to migrate any existing apps (ie, the migration of existing apps into the unified rules system will happen over the subsequent week or two). Not sure what the best way to communicate this is, but might be worth holding back from mentioning a specific date?

**Pull request merged**
